src/vm/dllimport.cpp

```diff
@@ -2950,6 +2960,236 @@ void PInvokeStaticSigInfo::DllImportInit(MethodDesc* pMD, LPCUTF8 *ppLibName, LP
+            pTarget >= gSysDllInfo[i].pImageBase &&
+            pTarget < gSysDllInfo[i].pImageBase + gSysDllInfo[i].dwImageSize)
+        {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+LPBYTE FollowIndirect(LPBYTE pTarget)
+{
+    CONTRACT(LPBYTE)
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        POSTCONDITION(CheckPointer(RETVAL, NULL_OK));
+    }
+    CONTRACT_END;
+
+    LPBYTE pRet = NULL;
+
+    EX_TRY
+    {
+        AVInRuntimeImplOkayHolder AVOkay;
+
+#ifdef _TARGET_X86_
+        if (pTarget != NULL && !(pTarget[0] != 0xff || pTarget[1] != 0x25))
+        {
+            pRet = **(LPBYTE**)(pTarget + 2);
+        }
+#elif defined(_TARGET_AMD64_)
+        if (pTarget != NULL && !(pTarget[0] != 0xff || pTarget[1] != 0x25))
+        {
+            INT64 rva = *(INT32*)(pTarget + 2);
+            pRet = *(LPBYTE*)(pTarget + 6 + rva);
+        }
+#endif
+    }
+        EX_CATCH
+    {
+        // Catch AVs here.
+    }
+    EX_END_CATCH(SwallowAllExceptions);
+
+    RETURN pRet;
+    }
+
+// This attempts to guess whether a target is an API call that uses SetLastError to communicate errors.
+BOOL HeuristicDoesThisLooksLikeAnApiCall(LPBYTE pTarget)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if (pTarget == NULL)
+        return FALSE;
+
+    if (HeuristicDoesThisLooksLikeAnApiCallHelper(pTarget))
+        return TRUE;
+
+    LPBYTE pTarget2 = FollowIndirect(pTarget);
+    if (pTarget2)
+    {
+        // jmp [xxxx] - could be an import thunk
+        return HeuristicDoesThisLooksLikeAnApiCallHelper(pTarget2);
+    }
+
+    return FALSE;
+}
+
+BOOL HeuristicDoesThisLookLikeAGetLastErrorCall(LPBYTE pTarget)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    static LPBYTE pGetLastError = NULL;
+    if (!pGetLastError)
+    {
+        // No need to use a holder here, since no cleanup is necessary.
+        HMODULE hMod = CLRGetModuleHandle(WINDOWS_KERNEL32_DLLNAME_W);
+        if (hMod)
+        {
+            pGetLastError = (LPBYTE)GetProcAddress(hMod, "GetLastError");
+            if (!pGetLastError)
+            {
+                // This should never happen but better to be cautious.
+                pGetLastError = (LPBYTE)-1;
+            }
+        }
+        else
+        {
+            // We failed to get the module handle for kernel32.dll. This is almost impossible
+            // however better to err on the side of caution.
+            pGetLastError = (LPBYTE)-1;
+        }
+    }
+
+    if (pTarget == pGetLastError)
+        return TRUE;
+
+    if (pTarget == NULL)
+        return FALSE;
+
+    LPBYTE pTarget2 = FollowIndirect(pTarget);
+    if (pTarget2)
+    {
+        // jmp [xxxx] - could be an import thunk
+        return pTarget2 == pGetLastError;
+    }
+
+    return FALSE;
+}
+
+DWORD __stdcall FalseGetLastError()
```

Comment:
The `__stdcall` can be removed. This is already the default when it matters.

Reply:
Please keep __stdcall. by default it is __cdcel  for C/C++.

Reply:
> by default it is __cdcel for C/C++.

It is not in our build on Windows. CoreCLR depends on being compiled with stdcall default on Windows in many places.

Reply:
@luqunl See below.

https://github.com/dotnet/coreclr/blob/04d9b557ef8b7c60a1194d062c7c2ee6e41fd8c1/configurecompiler.cmake#L509-L511

Reply:
> It is not in our build on Windows. CoreCLR depends on being compiled with stdcall default on Windows in many places.

Thanks.  Found out Build System will pass /Gz to cl.exe for x86 Windows.
BTW, There are a lot code in coreclr specified __stdcall.  Personally I would always recommend to specify calling convention.

Reply:
>There are a lot code in coreclr specified __stdcall. 

@luqunl That is true, but it is also the wrong thing to do. The reason it is done was because as you mentioned, `cl.exe` defaults to `cdecl`. Instead of ensuring the compiler is doing the right thing during build via flags people started adding the `__stdcall` which then requires making a #define for that when compiling in other platforms (see below). In almost all cases, the defaults should not be set so the code can look more natural the below list of nonsense can largely (not entirely) be avoided.

https://github.com/dotnet/coreclr/blob/04d9b557ef8b7c60a1194d062c7c2ee6e41fd8c1/src/pal/inc/pal_mstypes.h#L40-L66

Reply:
> I would always recommend to specify calling convention.

I have been recommending the exact opposite. The explicitly specified calling conventions are pain to deal with when porting to non-Windows platforms. The explicitly specified calling conventions have been deleted in many places e.g. https://github.com/dotnet/coreclr/commit/3108cc7e12e004e6c6bab2daf7826dd97957101f . The places that cannot live without explicitly specified calling convention have been switched to higher macros like `STDMETHODCALLTYPE` that can be defined to nothing everywhere but Windows x86.

Reply:
> then requires making a #define for that when compiling in other platforms (see below)

If coreclr product code need to use more than 1 call conventions, then the above file is necessary(at least __cdcel part).  

>> but it is also the wrong thing to do
if someone specified explicitly its calling convention, everyone can easily know what's calling convention for this native method instead of guess what is 'default' calling convention? whether compiler passing special flag to modify the default calling convention? 

>> so the code can look more natural the below list of nonsense can largely (not entirely) be avoided.

I would prefer add these 6 lines(__stdcall) code instead of  require us to not use __stdcall.

Reply:
>The places that cannot live without explicitly specified calling convention have been switched to higher macros like STDMETHODCALLTYPE that can be defined to nothing everywhere but Windows x86.

AFAIK, __stdcall can also be defined as nothing for platforms other than Windows x86.

Reply:
>AFAIK, __stdcall can also be defined as nothing for platforms other than Windows x86.

@luqunl Yes it can, but in MSVC it is a [keyword](https://docs.microsoft.com/en-us/cpp/cpp/stdcall) which implies semantic meaning that again is largely irrelevant when compiling on other platforms. Having a higher level macro is definitely preferred IMO, but only in cases where the concept is not default.

Reply:
Just so be clear another reason why the keyword usage is annoying is for IDEs and tooling in general. Defining these values when they macros or non-standard keywords plays hell with IDEs and make the tool writer's job even harder - having been one for a while I can attest to that. I don't know how many people use IDEs when looking at the code base, but it makes the language service's job much harder.

Reply:
Switched to ```STDMETHODCALLTYPE``` since it's reasonable to pick a convention on the interop boundary, but I don't want to cause cross-platform issues.